velox/functions/lib/TimeUtils.h

```diff
@@ -162,4 +162,151 @@ Timestamp truncateTimestamp(
     Timestamp timestamp,
     DateTimeUnit unit,
     const tz::TimeZone* timeZone);
+
+// Returns toTimestamp - fromTimestamp expressed in terms of unit.
+FOLLY_ALWAYS_INLINE int64_t diffTimestamp(
+    const DateTimeUnit unit,
+    const Timestamp& fromTimestamp,
+    const Timestamp& toTimestamp) {
+  // TODO(gaoge): Handle overflow and underflow with 64-bit representation
+  if (fromTimestamp == toTimestamp) {
+    return 0;
+  }
+
+  const int8_t sign = fromTimestamp < toTimestamp ? 1 : -1;
+
+  // fromTimepoint is less than or equal to toTimepoint
+  const std::chrono::
+      time_point<std::chrono::system_clock, std::chrono::milliseconds>
+          fromTimepoint(std::chrono::milliseconds(
+              std::min(fromTimestamp, toTimestamp).toMillis()));
+  const std::chrono::
+      time_point<std::chrono::system_clock, std::chrono::milliseconds>
+          toTimepoint(std::chrono::milliseconds(
+              std::max(fromTimestamp, toTimestamp).toMillis()));
+
+  // Millisecond, second, minute, hour and day have fixed conversion ratio
+  switch (unit) {
+    case DateTimeUnit::kMillisecond: {
+      return sign *
+          std::chrono::duration_cast<std::chrono::milliseconds>(
+              toTimepoint - fromTimepoint)
+              .count();
+    }
+    case DateTimeUnit::kSecond: {
+      return sign *
+          std::chrono::duration_cast<std::chrono::seconds>(
+              toTimepoint - fromTimepoint)
+              .count();
+    }
+    case DateTimeUnit::kMinute: {
+      return sign *
+          std::chrono::duration_cast<std::chrono::minutes>(
+              toTimepoint - fromTimepoint)
+              .count();
+    }
+    case DateTimeUnit::kHour: {
+      return sign *
+          std::chrono::duration_cast<std::chrono::hours>(
+              toTimepoint - fromTimepoint)
+              .count();
+    }
+    case DateTimeUnit::kDay: {
+      return sign *
+          std::chrono::duration_cast<date::days>(toTimepoint - fromTimepoint)
+              .count();
+    }
+    case DateTimeUnit::kWeek: {
+      return sign *
+          std::chrono::duration_cast<date::days>(toTimepoint - fromTimepoint)
+              .count() /
+          7;
+    }
+    default:
+      break;
+  }
+
+  // Spark support microsecond unit.
+  if (unit == DateTimeUnit::kMicrosecond) {
```

Comment:
Why not combine it to the switch case?

Reply:
Thank you, updated.

Reply:
Could you describe why it overflows? Presto UT should not test kMicrosecond

Reply:
Because you move change the toMills to toMicros, keep the toMills in L182, and combine them